core/src/main/java/lucee/runtime/db/QoQ.java

```diff
@@ -302,72 +543,132 @@ private Query getSingleTable(PageContext pc, Column table) throws PageException
+		if (exp instanceof Operation) return executeOperation(pc, sql, source, (Operation) exp, row);
+		if (exp instanceof BracketExpression) return executeBracked(pc, sql, source, (BracketExpression) exp, row);
+		throw new DatabaseException("unsupported sql statement [" + exp + "]", null, sql, null);
+	}
+
+	private Object executeExp(PageContext pc, SQL sql, Query source, Expression exp, int row, Object columnDefault) throws PageException {
+		if (exp instanceof Value) return ((Value) exp).getValue();// executeConstant(sql,qr,
+		if (exp instanceof Column) return executeColumn(pc, sql, source, (Column) exp, row, columnDefault);
+		if (exp instanceof Operation) return executeOperation(pc, sql, source, (Operation) exp, row);
+		if (exp instanceof BracketExpression) return executeBracked(pc, sql, source, (BracketExpression) exp, row);
 		throw new DatabaseException("unsupported sql statement [" + exp + "]", null, sql, null);
 	}
 
-	private Object executeExp(PageContext pc, SQL sql, Query qr, Expression exp, int row, Object columnDefault) throws PageException {
-		// print.o(exp.getClass().getName());
-		if (exp instanceof Value) return ((Value) exp).getValue();// executeConstant(sql,qr, (Value)exp, row);
-		if (exp instanceof Column) return executeColumn(sql, qr, (Column) exp, row, columnDefault);
-		if (exp instanceof Operation) return executeOperation(pc, sql, qr, (Operation) exp, row);
-		if (exp instanceof BracketExpression) return executeBracked(pc, sql, qr, (BracketExpression) exp, row);
+	/**
+	 * Accepts an expression which is the input to an aggregate operation.
+	 * 
+	 * @param pc
+	 * @param sql
+	 * @param source
+	 * @param exp
+	 * @param row
+	 * @param includeNull
+	 * @return an array with as many items as rows in the source query containing the corresponding
+	 *         expression result for each matching row
+	 * @throws PageException
+	 */
+	private Object[] executeAggregateExp(PageContext pc, SQL sql, Query source, Expression exp, boolean includeNull, boolean returnDistinct) throws PageException {
+		Object[] result = new Object[source.getRecordcount()];
+
+		// For a literal value, just fill an array with that value
+		if (exp instanceof Value) {
+			Object value = ((Value) exp).getValue();
+			for (int i = 0; i < source.getRecordcount(); i++) {
+				result[i] = value;
+			}
+			return result;
+		}
+
+		// For a column, return the data in that column as an array
+		if (exp instanceof Column) {
+			result = ((QueryColumnImpl) source.getColumn(((Column) exp).getColumn())).toArray();
+			// Simple return if we want all values
+			if (!returnDistinct && includeNull) {
+				return result;
+			}
+			// if we want to filter nulls or distinct the array
+			else {
+				Stream<Object> resultStream = Arrays.stream(result);
+				if (!includeNull) resultStream = resultStream.filter(s -> (s != null));
+				if (returnDistinct) resultStream = resultStream.distinct();
+				return resultStream.toArray(Object[]::new);
```

Comment:
that syntax will not work with Java 8 and we still compile with Java 8 for backward compatibility, so better symply do
`resultStream.toArray()`

Reply:
@michaeloffner Java the Stream API and Lambdas are part of JDK 8.  Are you just saying the `Object[]::new` part is not Java 8 compat?  I can remove that to just use the `toArray()` method on the Stream.

Reply:
i'm always using different JVM to work and build, just be on the safe side, openJDK was not complaining, but Oracle did. I try to be as conservative as possible with new functionality, i only using it when i need it and if it does not break backward compatibility. But i think you know that game with CFML compatibility ;-).